Plugins/NativeEngine/Source/IndexBuffer.cpp

```diff
@@ -35,42 +35,63 @@ namespace Babylon
                 bgfx::destroy(m_handle);
             }
         }
-        m_bytes.reset();
+
+        m_bytes.clear();
+
         m_disposed = true;
     }
 
-    void IndexBuffer::Update(Napi::Env env, gsl::span<uint8_t> bytes, uint32_t startIndex)
+    void IndexBuffer::Update(const gsl::span<uint8_t> bytes, uint32_t startIndex)
     {
         if (!m_dynamic)
         {
-            throw Napi::Error::New(env, "Cannot update non-dynamic index buffer.");
+            throw std::runtime_error{"Cannot update non-dynamic index buffer"};
         }
 
         if (bgfx::isValid(m_dynamicHandle))
         {
-            // Buffer was already created, do a real update operation.
             bgfx::update(m_dynamicHandle, startIndex, bgfx::copy(bytes.data(), static_cast<uint32_t>(bytes.size())));
         }
         else
         {
-            // Buffer hasn't been finalized yet, all that's necessary is to swap out the bytes.
-            m_bytes = {bytes.data(), bytes.data() + bytes.size()};
+            const size_t byteStride = (m_flags & BGFX_BUFFER_INDEX32) ? 4 : 2;
+            const size_t byteOffset = startIndex * byteStride;
+
+            if (byteOffset + bytes.size() > m_bytes.size())
+            {
+                throw std::runtime_error{"Failed to update index buffer: buffer overflow"};
+            }
+
+            std::memcpy(m_bytes.data() + byteOffset, bytes.data(), bytes.size());
         }
     }
 
-    bool IndexBuffer::CreateHandle()
+    void IndexBuffer::Set(bgfx::Encoder* encoder, uint32_t firstIndex, uint32_t numIndices)
     {
-        if (bgfx::isValid(m_handle))
+        if (!m_built)
+        {
+            Build();
+            m_built = true;
+        }
+
+        if (m_dynamic)
+        {
+            encoder->setIndexBuffer(m_dynamicHandle, firstIndex, numIndices);
+        }
+        else
         {
-            return true;
+            encoder->setIndexBuffer(m_handle, firstIndex, numIndices);
         }
+    }
 
+    void IndexBuffer::Build()
+    {
         auto releaseFn = [](void*, void* userData) {
-            auto* bytes = reinterpret_cast<decltype(m_bytes)*>(userData);
-            bytes->reset();
+            delete reinterpret_cast<decltype(m_bytes)*>(userData);
```

Comment:
is `userData` a pointer allocated by the the std::vector?

Reply:
It's a newly allocated pointer to `std::vector`.

See below:
```c++
auto* bytesPtr = new decltype(m_bytes){std::move(m_bytes)};
```